WIP - Revamping web services

hypeJunction opened the pull request:
DO NOT MERGE - WORK IN PROGRESS

What this does:
- Starts moving web services to OO
- Removes eval() and parameter serialization in favor of calling callables
- Simplifies parameter validation
- Adds API versioning
- Changes how callback function get their arguments - api 2.0 just passes an associative array instead of each parameter being a separate argument
- [ ] deprecate removed function in 1.x
- [ ] review existing texts
- [ ] add PHPUnit(?) tests for new classes

hypeJunction committed:
feature(web_services): remove eval-based execution

Removes eval() magic and serialization from web services to allow for a
wider range of callables as method callback
Improves handling and validation of parameters

hypeJunction committed:
feature(web_services): transition web services to OO API

Removes global state variables and transitions registration, authentication
and execution of API methods to classes
Adds _elgg_ws_registry() to instantiate a new registry
Adds an $assoc option that calls callback function with a single arguments
that contains an associative array of all input values

hypeJunction commented:
@Elgg/owners Any suggestions on the how to implement the registry?

hypeJunction commented on mod/web_services/start.php:
```diff
@@ -1,338 +1,255 @@
+function list_all_apis() {
+
+	$methods = ApiRegistry::all();
+	ksort($methods);
```
Need to ksort() for each API version

mrclay commented on mod/web_services/classes/ApiMethod.php:
```diff
@@ -0,0 +1,219 @@
+	 * @access private
+	 */
+	protected function getParams() {
+		global $API_METHODS;
```
I thought this was going away?

mrclay commented on mod/web_services/classes/ApiRegistry.php:
```diff
@@ -0,0 +1,68 @@
+
+	const API_VERSION = '2.0';
+	
+	static $methods = array();
```
Make this whole class dynamic and make _elgg_ws_registry() return a singleton instance.

hypeJunction committed:
security(web_services): remove expired user tokens on daily basis

Registers a cron handler to remove expired user tokens on daily basis

hypeJunction committed:
perf(web_services): refactor api_user library into OO

Refactors API user library functions into OO

hypeJunction committed:
feature(web_services): refactor api client library into OO

BREAKING CHANGE

Refactors api client into OO. If you were using get_standard_api_key_array()
or serialize_api_headers(), you need to remove them from your code, or
implement them in your code

hypeJunction committed:
feature(web_services): refactor user tokens into OO

Refactors user token related API to OO

hypeJunction commented:
Hm, something is wrong with the diffs...
I am done for today. Let me know what your thoughts are on this @Elgg/owners 
If you think this is acceptable, I will work on the tests.

hypeJunction commented:
There are few more things I want to do:

- create shortcut routing ```/api/2.0/system.api.list``` with optional ```?format=json|xml```
- deprecate ```elgg_ws_expose_function()``` in favour of:

```php
elgg_ws_register_method('GET', 'system.api.list', array(
  'function' => 'list_all_apis',
  'parameters' => array(),
  'require_api_auth' => false,
  'api_version' => '2.0',
));
```
- make method params filterable with a hook

ewinslow commented:
Can we have folks pass class names instead of function names. That way they
can be lazy loaded and (eventually) dependency injected?

Also: why explicitly support versioning instead of just having folks do it
manually like elgg_ws_register_method("system.api.list.v2",...)?

Last question: Does this mean we're sticking with rpc instead of doing
something more RESTful?

On Sun, Jun 28, 2015, 7:12 AM Ismayil Khayredinov <notifications@github.com>
wrote:

> There are few more things I want to do:
>
>    - create shortcut routing /api/2.0/system.api.list with optional
>    ?format=json|xml
>    - deprecate elgg_ws_expose_function() in favour of:
>
> elgg_ws_register_method('GET', 'system.api.list', array(  'function' => 'list_all_apis',  'parameters' => array(),  'require_api_auth' => false,  'api_version' => '2.0',));
>
>
>    - make method params filterable with a hook
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/8610#issuecomment-116285734>.
>

hypeJunction commented:
> Can we have folks pass class names instead of function names. That way they
> can be lazy loaded and (eventually) dependency injected?
I imagine it should work with ```__invoke()```. Haven't tried.

> Also: why explicitly support versioning instead of just having folks do it
> manually like elgg_ws_register_method("system.api.list.v2",...)?
Look at Method::execute(). We handle parameters differently. Plus, it will smoothen the transition. Unless, we just want to completely break all existing exposed methods.

> Last question: Does this mean we're sticking with rpc instead of doing
> something more RESTful?
I think we can move to rest with the next iteration. Need to clean up what we have first. I think REST will be realistic when we have better route maps in place.

hypeJunction commented:
Are plugin simpletests running on CI? I would expect Travis to break by now.

ewinslow commented:
Guess not...

mrclay commented:
If we're going to switch to REST anyway, why force devs to rewrite code now? Or will that transition be BC, too?

ewinslow commented:
Also see this alternative approach that seems to be gaining some traction: https://community.elgg.org/discussion/view/2158924/could-this-be-the-future-of-elgg-data-access-and-ui-rendering. High-level summary: implement RESTful principles at the application level and just use HTTP as the transport.

ewinslow commented:
I'd prefer to do this in a BC way for 2.x rather than submitting it as a breaking change for 2.0 at this point.

hypeJunction commented:
I can make it mostly BC, by not passing default api version, so existing methods will continue working. If folks want to switch to assoc array, they can pass version number. I guess that also means we do not maintain api registry api version keys - just using registration order to see which one takes precedence

hypeJunction commented:
@mrclay, I was thinking rest should be implemented in parallel, not as a replacement ( new plugin perhaps). I think we can eventually transition to REST as well in a BC way - just need to figure out how we will map query params to URL segments.

hypeJunction commented:
One way to make things RESTful is to create JSON/XML views for all of our resource pages and enable token-based PAM globally. Clients can then use web services to generate an auth token (or use their API token) to just request existing page URLs.
We could expose an additional method to request action tokens, so POST calls can be made directly to /action endpoints.

hypeJunction committed:
security(web_services): API authentication required by default

BREAKING CHANGE

All exposed methods that do not set API auth requirement to false will now
require API authentication

hypeJunction committed:
feature(web_services): service handling is now done by resource views

BREAKING CHANGE

Service handlers can no longer be registered/unregistered.
If you were using custom service handlers, move them to the
corresponding resource views under /resources/services/<service>/<handler>.

This gets rid of service handling boilerplate, and routes
/services to resource views.

Conflicts:
	docs/guides/upgrading.rst

hypeJunction committed:
fix(web_services): rest handler no longer defaults to json format

BREAKING CHANGE

Rest API handle will no longer default viewtype to json if the format provided
in the request is an unregistered viewtype. Exception will be thrown.

Conflicts:
	docs/guides/upgrading.rst

hypeJunction committed:
feature(web_services): all libraries are avilable globally

BREAKING CHANGE

Remove loading of elgg:ws and elgg:ws:client from plugins. These libraries are not longer registered, and all functions are available in core

Conflicts:
	docs/guides/upgrading.rst

hypeJunction commented:
I will break this down into multiple PRs.
